### PR TITLE
create DB client in BE that supports all Knex configuration

### PR DESCRIPTION
**Pull request by ydayagi**
flpath-660
https://issues.redhat.com/browse/FLPATH-660

### CODE REVIEW

File: plugins/notifications-backend/src/service/db.ts [hunk 1]
```diff
@@ -6,16 +6,16 @@ import knex, { Knex } from 'knex';
...
-      port: dbConfig.getNumber('connection.port'),
-      user: dbConfig.getString('connection.user'),
-      password: dbConfig.getString('connection.password'),
-      database: 'backstage_plugin_notifications',
-    },
-  });
+  const database = dbConfig.getOptional('connection.database')
+    ? {}
+    : { connection: { database: 'backstage_plugin_notifications' } };
+  const knexConfig = require('lodash').merge(
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/db.ts]:
```
import merge from 'lodash/merge';
....
merge({}, foo, bar);
```

File: plugins/notifications-backend/src/service/db.ts [hunk 2]
```diff
@@ -6,16 +6,16 @@ import knex, { Knex } from 'knex';
...
+    {},
+    dbConfig.get(),
+    dbConfig.getOptional('knexConfig'),
+    database,
```
Comment (mareklibra) [inline plugins/notifications-backend/src/service/db.ts]: What about following simplification?
```
merge(
{ connection: { database: 'backstage_plugin_notifications' }},
 dbConfig.get(),
dbConfig.getOptional('knexConfig'),
)
```

and removing the `const database` at line 9?

Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: i think it is easier to read this way. but whatever u prefer

File: plugins/notifications-backend/src/service/db.ts (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: i took this example from the backstage code. there's a problem to import lodash

Comment (mareklibra) [reply plugins/notifications-backend/src/service/db.ts]: What kind of problem? It works on other places.

File: plugins/notifications-backend/src/service/db.ts (hunk 2 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: i can not do what u suggest. if 'conneciton.databse' is defined then i need to make sure it is not overridden.

File: plugins/notifications-backend/src/service/db.ts (hunk 1 above)
Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: done

File: plugins/notifications-backend/src/service/db.ts (hunk 2 above)
Comment (mareklibra) [reply plugins/notifications-backend/src/service/db.ts]: Why? The configuration should have priority over hardcoded defaults.

The `merge` puts right argument to the left one, overriding existing during "deep merge".
So if `dbConfig` contains `connection.database`, the default will be overridden by the value from configuration.

Comment (ydayagi) [reply plugins/notifications-backend/src/service/db.ts]: fixed

**Review by mareklibra — APPROVED** [review]

**Pull request merged**